src/sage/crypto/boolean_function.pyx

```diff
@@ -1367,8 +1367,14 @@ cdef class BooleanFunction(SageObject):
             True
         """
         self._clear_cache()
+
+        if self._hamming_weight != -1 and self(int(i)) != int(y) & 1:
+            if int(y) & 1:
+                self._hamming_weight += 1
+            else:
+                self._hamming_weight -= 1
+        
         bitset_set_to(self._truth_table, int(i), int(y)&1)
```

Comment:
This line can be moved above too.

Reply:
like this?
```python
if self._hamming_weight != -1 and self(int(i)) != int(y) & 1:
    if int(y) & 1:
        self._hamming_weight += 1
    else:
        self._hamming_weight -= 1
    bitset_set_to(self._truth_table, int(i), int(y)&1)
```

Reply:
Uhh not quite, you have to move the hamming weight check too. So
```python
if self(int(i)) != int(y) & 1:
    if self._hamming_weight != -1:
        if int(y) & 1:
            self._hamming_weight += 1
        else:
            self._hamming_weight -= 1
    bitset_set_to(self._truth_table, int(i), int(y)&1)
```

Reply:
Oh I see